### PR TITLE
feat(opsem): add shadowmem annotations for {reset,is}_modified

### PR DESCRIPTION
**Pull request by priyasiddharth**
TODO: add annotations for other intrinsics like getBase(), getSize() in next commit

### CODE REVIEW

**Comment by priyasiddharth** [comment]
@agurfinkel can you PTAL?

File: lib/seadsa/ShadowMem.cc [hunk 1]
```diff
@@ -1177,6 +1189,59 @@ void ShadowMemImpl::visitMemhavoc(CallSite &CS) {
...
 
+void ShadowMemImpl::visitIsModified(CallSite &CS) {
+  LOG("dsa.ismodified", errs() << "Visiting isModified \n");
+  auto &callInst = *CS.getInstruction();
+  /* isModified definition:
+     bool @sea_is_modified(i8*)
+     operand 0 is the pointer to be marked
+  */
+  auto &ptr = *CS.getArgOperand(0);
+  if (!m_graph->hasCell(ptr)) {
```
Comment (agurfinkel) [inline lib/seadsa/ShadowMem.cc:1200]: there might be a bitcast that needs to be removed. But you will see it in the first test.

File: lib/seadsa/ShadowMem.cc [hunk 2]
```diff
@@ -1177,6 +1189,59 @@ void ShadowMemImpl::visitMemhavoc(CallSite &CS) {
...
+  }
+  const dsa::Cell &cell = m_graph->getCell(ptr);
+  if (cell.isNull()) {
+    LOG("dsa.ismodified", errs() << "cell is null for: " << ptr << "\n");
+    return;
+  }
+
+  m_B->SetInsertPoint(&callInst);
+  CallInst &memUse =
+      mkShadowLoad(*m_B, cell, dsa::getTypeSizeInBytes(*ptr.getType(), *m_dl));
```
Comment (agurfinkel) [inline lib/seadsa/ShadowMem.cc:1212]: maybe change size to 1. It is not used, but still, the instruction does not really load that many bytes

File: lib/seadsa/ShadowMem.cc [hunk 3]
```diff
@@ -1177,6 +1189,59 @@ void ShadowMemImpl::visitMemhavoc(CallSite &CS) {
...
+    LOG("dsa.resetmodified", errs() << "no cell for: " << ptr << "\n");
+    return;
+  }
+  const dsa::Cell &cell = m_graph->getCell(ptr);
+  if (cell.isNull()) {
+    LOG("dsa.resetmodified", errs() << "cell is null for: " << ptr << "\n");
+    return;
+  }
+
+  if (cell.getOffset() == 0) {
```
Comment (agurfinkel) [inline lib/seadsa/ShadowMem.cc:1234]: do not need this `if`. I think you have copied this from `calloc` hack

File: lib/seadsa/ShadowMem.cc [hunk 4]
```diff
@@ -1177,6 +1189,59 @@ void ShadowMemImpl::visitMemhavoc(CallSite &CS) {
...
+    m_B->SetInsertPoint(&callInst);
+    CallInst &memDef = mkShadowStore(
+        *m_B, cell, dsa::AllocSiteInfo::getAllocSiteSize(callInst));
+    associateConcretePtr(memDef, callInst, &callInst);
+  } else {
+    errs() << "WARNING: skipping sea.reset_modified instrumentation since "
```
Comment (agurfinkel) [inline lib/seadsa/ShadowMem.cc:1240]: `offset == 0` means that pointer is to the beginning of allocated object. I don't think we restrict flag reset to beginning of an object. Even if we do, `offset == 0` implies that pointer is to an object, but not the other way around.

File: lib/seadsa/ShadowMem.cc (hunk 1 above)
Comment (priyasiddharth) [reply lib/seadsa/ShadowMem.cc:1200]: This seems to work as is in unit tests and verify-c-common also.

File: lib/seadsa/ShadowMem.cc (hunk 2 above)
Comment (priyasiddharth) [reply lib/seadsa/ShadowMem.cc:1212]: done

File: lib/seadsa/ShadowMem.cc (hunk 3 above)
Comment (priyasiddharth) [reply lib/seadsa/ShadowMem.cc]: Thanks. Had fixed it previously but was in a separate commit which i forgot to upload.

File: lib/seadsa/ShadowMem.cc (hunk 4 above)
Comment (priyasiddharth) [reply lib/seadsa/ShadowMem.cc]: Ack.

**Comment by priyasiddharth** [comment]
I don't have write access to sea-dsa... someone will have to pull on my behalf.

thanks

**Pull request merged**